### PR TITLE
add key reference for alias and handle pagination

### PR DESCRIPTION
**Pull request by vijtrip2**
Issue #, if available: https://github.com/aws-controllers-k8s/community/issues/1313

Description of changes:
* Add tests for updating TargetKeyID in Alias resource
* Add the KeyRef inside Alias resource
* Handle pagination using custom code when listing alias. I will do a fast followup to update code-generator to handle the pagination.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: templates/hooks/alias/sdk_read_many_pre_set_output.go.tpl [hunk 1]
```diff
@@ -1,12 +1,28 @@
...
+    for _, elem := range aliases {
+      if elem.AliasName == nil || r.ko.Spec.Name == nil {
+        continue
+      }
+
+      if *elem.AliasName == *r.ko.Spec.Name {
+        matchingAliases = append(matchingAliases, elem)
+      }
+    }
+    resp.Aliases = matchingAliases
```
Comment (jaypipes) [inline templates/hooks/alias/sdk_read_many_pre_set_output.go.tpl:28]: nit... clean up the indentation in this file :) looks like all tabs are replaced with spaces?

File: templates/hooks/alias/sdk_update_post_build_request.go.tpl [hunk 2]
```diff
@@ -0,0 +1 @@
+    input.AliasName = desired.ko.Spec.Name
```
Comment (jaypipes) [inline templates/hooks/alias/sdk_update_post_build_request.go.tpl:1]: So, unless I'm mistaken, by setting this, you basically guarantee that you won't need to paginate through more than a single result, no?

File: test/e2e/tests/test_alias.py [hunk 3]
```diff
@@ -14,17 +14,27 @@
...
+from e2e.tests.helper import KMSValidator
 
+CREATE_WAIT_AFTER_SECONDS = 30
+MODIFY_WAIT_AFTER_SECONDS = 30
 DELETE_WAIT_AFTER_SECONDS = 30
+DELETE_WAIT_PERIODS = 3
+DELETE_WAIT_PERIOD_LENGTH_SECONDS = 10
 
-KEY_RESOURCE_PLURAL = "aliases"
+KEY_RESOURCE_PLURAL = "keys"
```
Comment (RedbackThomson) [inline test/e2e/tests/test_alias.py:34]: import from the key test file instead?

File: test/e2e/tests/test_alias.py [hunk 4]
```diff
@@ -63,25 +81,107 @@ def simple_alias(simple_key):
...
+
+        # Create Alias. Needs Key reference
+        k8s.create_custom_resource(alias_ref, alias_res_data)
+
+        # Create Key. Needs no reference
+        k8s.create_custom_resource(key_ref, key_res_data)
+
+        time.sleep(CREATE_WAIT_AFTER_SECONDS)
+
+        assert k8s.wait_on_condition(key_ref, "ACK.ResourceSynced", "True", wait_periods=10)
```
Comment (RedbackThomson) [inline test/e2e/tests/test_alias.py:163]: Replace the string const with an imported condition variable: https://github.com/aws-controllers-k8s/test-infra/blob/main/src/acktest/k8s/condition.py#L21

File: templates/hooks/alias/sdk_read_many_post_build_request.go.tpl [hunk 5]
```diff
@@ -1,2 +0,0 @@
-// Match by associated Key ID
-input.KeyId = r.ko.Spec.TargetKeyID
```
Comment (jaypipes) [inline templates/hooks/alias/sdk_read_many_post_build_request.go.tpl:2]: Why was this removed, though?

Comment (jaypipes) [reply templates/hooks/alias/sdk_read_many_post_build_request.go.tpl:2]: The reason I ask is because this was previously avoiding the need to paginate...

Comment (RedbackThomson) [reply templates/hooks/alias/sdk_read_many_post_build_request.go.tpl:2]: That's what I thought too, but not entirely. There can be multiple aliases for any 1 key. This just filters the number of returned aliases down. There still may be too many to return in a single response - so pagination is still required.

Comment (RedbackThomson) [reply templates/hooks/alias/sdk_read_many_post_build_request.go.tpl:2]: Oh `ListAliases` defaults to 50 aliases returned. So there is no need for pagination here. https://docs.aws.amazon.com/kms/latest/APIReference/API_ListAliases.html

Even if the maximum # aliases per key was higher, the API can return up to 100 aliases.

Comment (vijtrip2) [reply templates/hooks/alias/sdk_read_many_post_build_request.go.tpl:2]: >So there is no need for pagination here.

Per Key filter returns NotFound for an existing alias which is not correct and does not allow update path to be triggered.

And when we are not listing per key, there can be many aliases and then pagination will help avoid false positive NotFound errors too.

Comment (jaypipes) [reply templates/hooks/alias/sdk_read_many_post_build_request.go.tpl:2]: > > So there is no need for pagination here.
> 
> Per Key filter returns NotFound for an existing alias(during update) which is not correct and does not allow update path to be triggered.
> 
> And when we are not listing per key, there can be many aliases and then pagination will help avoid false positive NotFound errors too.

OK, that's just weird :)
